README.md
<!-- Improved compatibility of back to top link: See: https://github.com/othneildrew/Best-README-Template/pull/73 -->

<a name="readme-top"></a>

[![NPM](https://img.shields.io/npm/v/erc721l?color=%23cc3534&style=for-the-badge&logo=npm&logoColor=white)](https://www.npmjs.com/package/erc721l)
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/Racks-Labs/ERC721-Linkable">
    <picture>
      <source media="(prefers-color-scheme: light)" srcset="README/E7L-Black.svg">
      <source media="(prefers-color-scheme: dark)" srcset="README/E7L-White.svg">
      <img src="README/E7L-Black.svg" alt="E7L logo" width="500">
    </picture>
  </a>
<h3 align="center">E7L Standard</h3>
  <p align="center">
    <b>A new way of linking 🔗</b>
    <br />
    <a href="https://e7l.rackslabs.com"><strong>Explore the website »</strong></a>
    <br />
    <br />
    <a href="https://github.com/Racks-Labs/ERC721-Linkable/issues">Report Bug</a>
    ·
    <a href="https://github.com/Racks-Labs/ERC721-Linkable/issues">Request Feature</a>
  </p>
</div>

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-erc-721-linkable">About ERC 721-Linkable</a>
    </li>
    <li><a href="#technical-specs">Technical specs</a></li>
    <li><a href="#getting-started">Getting Started</a></li>
    <li><a href="#future-improvements">Future improvements</a></li>
  </ol>
</details>

## About ERC 721 Linkable

[![E7L teaser][e7l-teaser]](https://e7l.rackslabs.com)

The goal of ERC-721Linkable(E7L) is to allow nfts to be linked. The nfts that implements this standard is non-transferable until it is linked to a nft of other smart contract (parent token), once the E7L is linked it can only be transferred to the current owner of the parent token.

It creates a bunch of new opportunities to bring value to you holders and create new online experiences. Let's see 2 possible use cases below.

E7L allows you to create experiences where people can earn an E7L and link it to their favourite nft of your project increasing the perceived value of it and allowing an NFT to have on chain reputation, kind of POAP but for tokens.

It can also be used to create a multi nft drop, for example a physically backed hoodie that has a linked pfp, a decentralized version of that hoodie and a nft 3d render to show on oncyber. The ownership of all these nfts will be of the current owner of the hoodie pbt token automatically, and if that token is transferred all these complementary nft can be claim by the new owner.

## Technical specs

ERC-721Linkable is an extension of the ERC-721, the main technical add-ons are the variable parentContract, the struct LinkableToken, and the `_linkToken` `_unlinkToken` and `syncToken` functions.

The function `_linkToken` initialize a minted token id linking it to a tokenId of the parent contract. If a token is not initialized it can not be transferred, once it is initialized it can only be transferred using the `syncToken` function that will transfer the nft to the current owner of the parent token ID. The function `_unlinkToken` reverts that token to the initial state. This last function is intended to be used only with an access control management.

## Getting started

In order to use E7L you just need to install the following npm package.

```shell
npm i erc721l
```

or

```shell
yarn add erc721l
```

<br/>

Once you have all the contracts you just need to create a new solidity file, import the ERC721Linkable contract and make your contract.

```solidity
pragma solidity ^0.8.7;

import "erc721l/contracts/ERC721Linkable.sol";

contract E7L is ERC721Linkable {
  constructor(
    string memory _name,
    string memory _symbol,
    address _parentContract
  ) ERC721Linkable(_name, _symbol, _parentContract) {}

  function mint(uint256 tokenId) public {
    _safeMint(msg.sender, tokenId);
  }

  // CUSTOM CODE GOES HERE
}
```

## Future improvements

During the development process of the standard we have realized that the ui and ux is really important in a project that implements this standard. So we are working on a sub graph base repo to help with that.

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->

[contributors-shield]: https://img.shields.io/github/contributors/Racks-Labs/ERC721-Linkable.svg?style=for-the-badge
[contributors-url]: https://github.com/Racks-Labs/ERC721-Linkable/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/Racks-Labs/ERC721-Linkable.svg?style=for-the-badge
[forks-url]: https://github.com/Racks-Labs/ERC721-Linkable/network/members
[stars-shield]: https://img.shields.io/github/stars/Racks-Labs/ERC721-Linkable.svg?style=for-the-badge
[stars-url]: https://github.com/Racks-Labs/ERC721-Linkable/stargazers
[issues-shield]: https://img.shields.io/github/issues/Racks-Labs/ERC721-Linkable.svg?style=for-the-badge
[issues-url]: https://github.com/Racks-Labs/ERC721-Linkable/issues
[e7l-teaser]: README/Teaser.gif


CHANGELOG.md
# erc721l

## 1.0.1

### Patch Changes

- 3421fc9: udpate depencies and refactor E7L standard


test/Basic.ts
import { expect } from "chai";
import { E7LBasic, MRCRYPTO } from "../typechain-types";
import { deployBasic } from "../utils/deployBasic";
import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
import { Signer } from "ethers";
import { ethers } from "hardhat";

describe("E7L: Basic functionality test", function () {
  let E7L: E7LBasic, MRC: MRCRYPTO;
  let jommys: Signer, yonathan: Signer;

  async function deploy() {
    const deployed = await deployBasic();
    E7L = deployed.E7L;
    MRC = deployed.MRC;
    yonathan = deployed.yonathan;
    jommys = deployed.jommys;
  }

  beforeEach(async function () {
    await loadFixture(deploy);
  });

  it("Check ownership of token 0", async function () {
    expect(await E7L.ownerOf(0)).to.be.equal(await yonathan.getAddress());
    expect(await E7L.balanceOf(await yonathan.getAddress())).to.be.equal(1);
  });

  describe("linkToken()", function () {
    it("Should not be linked", async function () {
      const res = await E7L.tokenInfo(0);
      expect(res.parentContract).to.be.equal(ethers.ZeroAddress);
      expect(res.parentTokenId).to.be.equal(0);
    });

    it("Should link token", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());

      const res = await E7L.tokenInfo(0);

      expect(res.parentContract).to.be.not.equal(ethers.ZeroAddress);
      expect(res.parentTokenId).to.be.equal(2);
    });
  });
  describe("unlinkToken()", function () {
    it("Should unlink token", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());

      let res = await E7L.tokenInfo(0);

      expect(res.parentContract).to.be.equal(await MRC.getAddress());
      expect(res.parentTokenId).to.be.equal(2);
      await E7L.connect(yonathan).unlinkToken(0);

      res = await E7L.tokenInfo(0);

      expect(res.parentContract).to.be.equal(ethers.ZeroAddress);
      expect(res.parentTokenId).to.be.equal(0);
    });
  });
  describe("syncToken()", function () {
    it("Should not transfer token", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());
      await MRC.connect(yonathan).transferFrom(
        yonathan.getAddress(),
        jommys.getAddress(),
        2,
      );
      expect(await MRC.ownerOf(2)).to.be.equal(await jommys.getAddress());
      expect(await E7L.ownerOf(0)).to.be.equal(await yonathan.getAddress());
    });

    it("Should transfer token", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());
      await MRC.connect(yonathan).transferFrom(
        yonathan.getAddress(),
        jommys.getAddress(),
        2,
      );
      expect(await MRC.ownerOf(2)).to.be.equal(await jommys.getAddress());
      expect(await E7L.ownerOf(0)).to.be.equal(await yonathan.getAddress());
      await E7L.syncToken(0);
      expect(await E7L.ownerOf(0)).to.be.equal(await jommys.getAddress());
    });
  });
});


test/Reverts.ts
import { expect } from "chai";
import { E7LBasic, MRCRYPTO } from "../typechain-types";
import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
import { Signer } from "ethers";

import { deployBasic } from "../utils/deployBasic";

describe("E7L: Reverts test", function () {
  let E7L: E7LBasic, MRC: MRCRYPTO;
  let jommys: Signer, yonathan: Signer;

  async function deploy() {
    const deployed = await deployBasic();

    E7L = deployed.E7L;
    MRC = deployed.MRC;
    yonathan = deployed.yonathan;
    jommys = deployed.jommys;
  }

  beforeEach(async function () {
    await loadFixture(deploy);
  });

  it("Check ownership of token 0", async function () {
    expect(await E7L.ownerOf(0)).to.be.equal(await yonathan.getAddress());
    expect(await E7L.balanceOf(await yonathan.getAddress())).to.be.equal(1);
  });

  describe("_beforeTokenTransfer()", function () {
    it("Should revert with invalid address", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());
      await expect(
        E7L.connect(yonathan).transferFrom(
          yonathan.getAddress(),
          jommys.getAddress(),
          0,
        ),
      ).to.be.revertedWith(
        "ERC721Linkable: the 'to' address is not the legitimate owner",
      );
    });
  });
  describe("linkToken()", function () {
    it("Should revert with invalid token ID", async function () {
      await expect(
        E7L.connect(yonathan).linkToken(100, 1, MRC.getAddress()),
      ).to.be.revertedWith("ERC721: invalid token ID");
      await expect(
        E7L.connect(yonathan).linkToken(0, 100000, MRC.getAddress()),
      ).to.be.revertedWith("ERC721: owner query for nonexistent token");
    });

    it("Should revert with token already linked", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());
      await expect(
        E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress()),
      ).to.be.revertedWith("ERC721Linkable: token is already linked");
    });

    it("Should revert with caller is not owner nor aproved", async function () {
      await expect(
        E7L.connect(jommys).linkToken(0, 2, MRC.getAddress()),
      ).to.be.revertedWith("ERC721: caller is not token owner nor approved");
    });
  });
  describe("syncToken()", function () {
    it("Should revert with token already synced", async function () {
      await E7L.connect(yonathan).linkToken(0, 2, MRC.getAddress());
      await expect(E7L.syncToken(0)).to.be.revertedWith(
        "ERC721Linkable: token already synced",
      );
    });
  });
  describe("tokenInfo()", function () {
    it("Should revert with invalid token ID", async function () {
      await expect(E7L.tokenInfo(3)).to.be.revertedWith(
        "ERC721: invalid token ID",
      );
    });
  });
});


test/E7LManager.ts
import { expect } from "chai";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
import { E7LBasic, E7LManager, MRCRYPTO } from "../typechain-types";
import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
import { deployE7LManager } from "../utils/deployE7LManager";

describe("E7LManager tests", function () {
  let E7L: E7LBasic, MRC: MRCRYPTO, E7LManager: E7LManager;
  let tokens: Array<E7LManager.LinkedTokenStruct>;
  let jommys: SignerWithAddress, yonathan: SignerWithAddress;

  const MAX_BATCH_NUMBER = 150;

  async function deploy() {
    const deployE7LM = await deployE7LManager();
    E7L = deployE7LM.E7L;
    MRC = deployE7LM.MRC;
    E7LManager = deployE7LM.E7LManager;
    yonathan = deployE7LM.yonathan;
    jommys = deployE7LM.jommys;
  }

  beforeEach(async function () {
    await loadFixture(deploy);

    tokens = [
      { id: 0, contractAddress: E7L.address },
      { id: 1, contractAddress: E7L.address },
      { id: 2, contractAddress: E7L.address },
    ];
  });

  it("Check ownership of tokens 0, 1, 2", async function () {
    expect(await E7L.ownerOf(0)).to.be.equal(yonathan.address);
    expect(await E7L.ownerOf(1)).to.be.equal(yonathan.address);
    expect(await E7L.ownerOf(2)).to.be.equal(yonathan.address);
    expect(await E7L.balanceOf(yonathan.address)).to.be.equal(3);
  });

  describe("linkTokens()", function () {
    it("Should not have any linked token", async function () {
      expect(
        (await E7LManager.getLinkedTokensFromParent(MRC.address, 2)).length,
      ).to.be.equal(0);
    });

    it("Should link 3 tokens", async function () {
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);

      const res = await E7L.tokenInfo(0);
      expect(res.linked).to.be.true;
      expect(res.parentTokenId).to.be.equal(2);

      const res1 = await E7L.tokenInfo(1);
      expect(res1.linked).to.be.true;
      expect(res1.parentTokenId).to.be.equal(2);

      const res2 = await E7L.tokenInfo(2);
      expect(res2.linked).to.be.true;
      expect(res2.parentTokenId).to.be.equal(2);
    });

    it("Should link MAX_BATCH_NUMBER tokens", async function () {
      for (let i = 3; i < MAX_BATCH_NUMBER; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);
      expect(
        (await E7LManager.getLinkedTokensFromParent(MRC.address, 2)).length,
      ).to.be.equal(MAX_BATCH_NUMBER);
    });

    it("Should link tokens in several batchesShould link tokens in several batches", async function () {
      for (let i = 3; i < MAX_BATCH_NUMBER; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);

      const tokens2 = [];
      await E7L.connect(yonathan).setApprovalForAll(E7LManager.address, true);

      for (let i = MAX_BATCH_NUMBER; i < MAX_BATCH_NUMBER * 2; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens2.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens2);

      expect(
        (await E7LManager.getLinkedTokensFromParent(MRC.address, 2)).length,
      ).to.be.equal(MAX_BATCH_NUMBER * 2);

      const res = await E7L.tokenInfo(MAX_BATCH_NUMBER - 1);
      expect(res.linked).to.be.true;
      expect(res.parentTokenId).to.be.equal(2);
    });
  });
  describe("syncTokens()", function () {
    it("Should not transfer tokens", async function () {
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);
      await MRC.connect(yonathan).transferFrom(
        yonathan.address,
        jommys.address,
        2,
      );
      expect(await MRC.ownerOf(2)).to.be.equal(jommys.address);
      expect(await E7L.ownerOf(0)).to.be.equal(yonathan.address);
    });

    it("Should transfer 3 tokens", async function () {
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);
      await MRC.connect(yonathan).transferFrom(
        yonathan.address,
        jommys.address,
        2,
      );
      expect(await MRC.ownerOf(2)).to.be.equal(jommys.address);
      await E7LManager.syncTokens(MRC.address, 2);

      expect(await E7L.ownerOf(0)).to.be.equal(jommys.address);
      expect(await E7L.ownerOf(1)).to.be.equal(jommys.address);
      expect(await E7L.ownerOf(2)).to.be.equal(jommys.address);
    });

    it("Should transfer MAX_BATCH_NUMBER tokens", async function () {
      for (let i = 3; i < MAX_BATCH_NUMBER; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);
      await MRC.connect(yonathan).transferFrom(
        yonathan.address,
        jommys.address,
        2,
      );

      expect(await MRC.ownerOf(2)).to.be.equal(jommys.address);
      await E7LManager.syncTokens(MRC.address, 2);
      for (let i = 0; i < MAX_BATCH_NUMBER; i++) {
        expect(await E7L.ownerOf(i)).to.be.equal(jommys.address);
      }
    });

    it("Should transfer above MAX_BATCH_NUMBER tokens", async function () {
      for (let i = 3; i < MAX_BATCH_NUMBER; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens);

      const tokens2 = [];
      for (let i = MAX_BATCH_NUMBER; i < MAX_BATCH_NUMBER * 2; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens2.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens2);

      const tokens3 = [];
      for (let i = MAX_BATCH_NUMBER * 2; i < MAX_BATCH_NUMBER * 3; i++) {
        await E7L.connect(yonathan).mint(i);
        tokens3.push({ id: i, contractAddress: E7L.address });
      }
      await E7LManager.connect(yonathan).linkTokens(MRC.address, 2, tokens3);

      await MRC.connect(yonathan).transferFrom(
        yonathan.address,
        jommys.address,
        2,
      );

      expect(await MRC.ownerOf(2)).to.be.equal(jommys.address);
      await E7LManager.syncTokens(MRC.address, 2);
      for (let i = 0; i < MAX_BATCH_NUMBER * 3; i++) {
        expect(await E7L.ownerOf(i)).to.be.equal(jommys.address);
      }
    });
  });
});


utils/deploy.ts
import { ethers } from "hardhat";
import { env } from "../env";
import { E7LBasic, MRCRYPTO } from "../typechain-types";
import reset from "./reset";

const MR_CRYPTO_ADDRESS = "0xeF453154766505FEB9dBF0a58E6990fd6eB66969";
const YONATHAN_ADDRESS = "0x4C9a3E12e523493383dd59162ECc8a26812192bE";
const JOMMYS_ADDRESS = "0x0AeaC6D1424EA6d0F87123A50CA5eEc9f16108c5";

export async function deployContracts() {
  const yonathan = await ethers.getImpersonatedSigner(YONATHAN_ADDRESS);
  const jommys = await ethers.getImpersonatedSigner(JOMMYS_ADDRESS);

  let MRC: MRCRYPTO, E7L: E7LBasic;

  if (env.TEST_LOCAL_BLOCKCHAIN) {
    // Deploy a Mock of My.Crypto Contract
    const MRC_Factory = await ethers.getContractFactory("MRCRYPTO");
    MRC = await MRC_Factory.deploy(
      "MrCryptoMock",
      "MRC",
      "https://apinft.racksmafia.com/api/",
      "https://apinft.racksmafia.com/api/hidden.json",
    );

    await MRC.deployed();

    console.log("MRC deployed to:", MRC.address);

    // Deploy a Mock of E7L Contract
    const E7L_Factory = await ethers.getContractFactory("E7LBasic");
    E7L = await E7L_Factory.deploy("E7L", "E7L", MRC.address);

    await E7L.deployed();

    console.log("E7L deployed to:", E7L.address);

    const [owner] = await ethers.getSigners();

    await owner.sendTransaction({
      to: YONATHAN_ADDRESS,
      value: ethers.utils.parseEther("10.0"),
    });

    await owner.sendTransaction({
      to: JOMMYS_ADDRESS,
      value: ethers.utils.parseEther("10.0"),
    });

    await MRC.mint(1);
    // Mint the Mr.Crypto #2 for Yonathan
    await MRC.connect(yonathan).mint(1);
    // Mint the Mr.Crypto #3 for Yonathan
    await MRC.connect(jommys).mint(1);
  } else {
    // Use de actual Mr.Crypto Contract deployed on polygon with hardhat fork
    await reset();

    // Is necessary to do this after reset the fork
    await ethers.getImpersonatedSigner(YONATHAN_ADDRESS);
    await ethers.getImpersonatedSigner(JOMMYS_ADDRESS);

    MRC = await ethers.getContractAt("MRCRYPTO", MR_CRYPTO_ADDRESS);

    const E7L_Factory = await ethers.getContractFactory("E7LBasic");
    E7L = await E7L_Factory.connect(jommys).deploy("E7L", "E7L", MRC.address);
  }

  await E7L.connect(yonathan).mint(0);

  return { MRC, E7L, yonathan, jommys };
}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
deployContracts().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});


utils/upgradeable/deployUpgradeable.ts
import { ethers } from "hardhat";
import { env } from "../../env";
import { E7LUpgradeableBasic, MRCRYPTO } from "../../typechain-types";
import reset from "../reset";

const MR_CRYPTO_ADDRESS = "0xeF453154766505FEB9dBF0a58E6990fd6eB66969";
const YONATHAN_ADDRESS = "0x4C9a3E12e523493383dd59162ECc8a26812192bE";
const JOMMYS_ADDRESS = "0x0AeaC6D1424EA6d0F87123A50CA5eEc9f16108c5";

export async function deployBasic() {
  const yonathan = await ethers.getImpersonatedSigner(YONATHAN_ADDRESS);
  const jommys = await ethers.getImpersonatedSigner(JOMMYS_ADDRESS);

  let MRC: MRCRYPTO, E7L_Proxy: E7LUpgradeableBasic;

  if (env.TEST_LOCAL_BLOCKCHAIN) {
    // Deploy a Mock of My.Crypto Contract
    const MRC_Factory = await ethers.getContractFactory("MRCRYPTO");
    MRC = await MRC_Factory.deploy(
      "MrCryptoMock",
      "MRC",
      "https://apinft.racksmafia.com/api/",
      "https://apinft.racksmafia.com/api/hidden.json",
    );

    await MRC.waitForDeployment();

    // Deploy a Mock of E7L Contract
    const E7L_Factory = await ethers.getContractFactory("E7LUpgradeableBasic");
    const E7L = await E7L_Factory.deploy();
    await E7L.waitForDeployment();

    const E7LProxy_Factory = await ethers.getContractFactory("E7LProxy");
    const E7LProxy_Raw = await E7LProxy_Factory.deploy(E7L.getAddress());
    await E7LProxy_Raw.waitForDeployment();

    E7L_Proxy = E7L.attach(
      await E7LProxy_Raw.getAddress(),
    ) as E7LUpgradeableBasic;

    const tx = await E7L_Proxy.initialize("E7L", "E7L", MRC.getAddress());
    await tx.wait();

    // Send ETH to Yonathan and Jommys
    const [owner] = await ethers.getSigners();

    await owner.sendTransaction({
      to: YONATHAN_ADDRESS,
      value: ethers.parseEther("100.0"),
    });

    await owner.sendTransaction({
      to: JOMMYS_ADDRESS,
      value: ethers.parseEther("100.0"),
    });

    await MRC.mint(1);
    // Mint the Mr.Crypto #2 for Yonathan
    await MRC.connect(yonathan).mint(1);
    // Mint the Mr.Crypto #3 for Jommys
    await MRC.connect(jommys).mint(1);
  } else {
    // Use de actual Mr.Crypto Contract deployed on polygon with hardhat fork
    await reset();

    // Is necessary to do this after reset the fork to use Jommys' and Yonathan's accounts
    await ethers.getImpersonatedSigner(YONATHAN_ADDRESS);
    await ethers.getImpersonatedSigner(JOMMYS_ADDRESS);

    MRC = await ethers.getContractAt("MRCRYPTO", MR_CRYPTO_ADDRESS);

    const E7L_Factory = await ethers.getContractFactory("E7LUpgradeableBasic");
    const E7L = await E7L_Factory.connect(jommys).deploy();
    await E7L.waitForDeployment();

    const E7LProxy_Factory = await ethers.getContractFactory("E7LProxy");
    const E7LProxy_Raw = await E7LProxy_Factory.deploy(E7L.getAddress());
    await E7LProxy_Raw.waitForDeployment();

    E7L_Proxy = E7L.attach(
      await E7LProxy_Raw.getAddress(),
    ) as E7LUpgradeableBasic;

    const tx = await E7L_Proxy.initialize("E7L", "E7L", MRC.getAddress());
    await tx.wait();
  }

  // Mint the E7L with id 0 for Yonathan
  await E7L_Proxy.connect(yonathan).mint(0);

  return { MRC, E7L_Proxy, yonathan, jommys };
}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
deployBasic().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});


utils/deployBasic.ts
import { ethers } from "hardhat";
import { env } from "../env";
import { E7LBasic, MRCRYPTO } from "../typechain-types";
import reset from "./reset";

const MR_CRYPTO_ADDRESS = "0xeF453154766505FEB9dBF0a58E6990fd6eB66969";
const YONATHAN_ADDRESS = "0x4C9a3E12e523493383dd59162ECc8a26812192bE";
const JOMMYS_ADDRESS = "0x0AeaC6D1424EA6d0F87123A50CA5eEc9f16108c5";

export async function deployBasic() {
  const yonathan = await ethers.getImpersonatedSigner(YONATHAN_ADDRESS);
  const jommys = await ethers.getImpersonatedSigner(JOMMYS_ADDRESS);

  let MRC: MRCRYPTO, E7L: E7LBasic;

  if (env.TEST_LOCAL_BLOCKCHAIN) {
    // Deploy a Mock of My.Crypto Contract
    const MRC_Factory = await ethers.getContractFactory("MRCRYPTO");
    MRC = await MRC_Factory.deploy(
      "MrCryptoMock",
      "MRC",
      "https://apinft.racksmafia.com/api/",
      "https://apinft.racksmafia.com/api/hidden.json",
    );

    await MRC.waitForDeployment();

    // Deploy a Mock of E7L Contract
    const E7L_Factory = await ethers.getContractFactory("E7LBasic");
    E7L = await E7L_Factory.deploy("E7L", "E7L", MRC.getAddress());

    await E7L.waitForDeployment();

    // Send ETH to Yonathan and Jommys
    const [owner] = await ethers.getSigners();

    await owner.sendTransaction({
      to: YONATHAN_ADDRESS,
      value: ethers.parseEther("100.0"),
    });

    await owner.sendTransaction({
      to: JOMMYS_ADDRESS,
      value: ethers.parseEther("100.0"),
    });

    await MRC.mint(1);
    // Mint the Mr.Crypto #2 for Yonathan
    await MRC.connect(yonathan).mint(1);
    // Mint the Mr.Crypto #3 for Jommys
    await MRC.connect(jommys).mint(1);
  } else {
    // Use de actual Mr.Crypto Contract deployed on polygon with hardhat fork
    await reset();

    // Is necessary to do this after reset the fork to use Jommys' and Yonathan's accounts
    await ethers.getImpersonatedSigner(YONATHAN_ADDRESS);
    await ethers.getImpersonatedSigner(JOMMYS_ADDRESS);

    MRC = await ethers.getContractAt("MRCRYPTO", MR_CRYPTO_ADDRESS);

    const E7L_Factory = await ethers.getContractFactory("E7LBasic");
    E7L = await E7L_Factory.connect(jommys).deploy(
      "E7L",
      "E7L",
      MRC.getAddress(),
    );
  }

  // Mint the E7L with id 0 for Yonathan
  await E7L.connect(yonathan).mint(0);

  return { MRC, E7L, yonathan, jommys };
}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
deployBasic().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
